README.md
<div align="center">
	<br>
	<br>
	<a href="https://foodsharing.de">
		<!-- image is docs/images/FS_Logo_gb_RGB.png but hosted publicly -->
		<img src="https://user-images.githubusercontent.com/31616/42413241-8802b03c-821c-11e8-91c5-f94930313290.png" alt="foodsharing">
	</a>
	<br>
	<br>
</div>

<!-- image is docs/images/fsscreenshot.png but hosted pubicly -->
![](https://user-images.githubusercontent.com/31616/42418486-f8f571e6-82a1-11e8-8771-41e403944101.png)

## 🌍 Overview

This is the code that powers
[foodsharing.de](https://foodsharing.de), 
[foodsharing.at](https://foodsharing.at), and
[foodsharingschweiz.ch](https://foodsharingschweiz.ch).

## 💻 Development

Visit [devdocs.foodsharing.network](https://devdocs.foodsharing.network) to get started with development.

## 💒 Community

Our developers hang out in the
[yunity slack](https://slackin.yunity.org) group in the
**#foodsharing-dev** and **#foodsharing-beta** channels. Come and say hi!

## 🔣 Translations

Translations are done via 
[weblate](https://hosted.weblate.org/projects/foodsharing/). If you can help translate the app into a
new language, feel free to help!

## 👪 Contributors

<!-- FOODSHARING-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<span>💻&nbsp;=&nbsp;Code</span>&nbsp;&nbsp;<span>📝&nbsp;=&nbsp;Documentation</span>&nbsp;&nbsp;<span>🎨&nbsp;=&nbsp;Design</span>&nbsp;&nbsp;<span>🐜&nbsp;=&nbsp;Bug reports</span>&nbsp;&nbsp;<span>💡&nbsp;=&nbsp;Ideas, Planning, & Feedback</span>&nbsp;&nbsp;<span>🔐&nbsp;=&nbsp;Security</span>&nbsp;&nbsp;<span>🔩&nbsp;=&nbsp;Infrastructure (Hosting, Build-Tools, etc)</span>&nbsp;&nbsp;<span>👀&nbsp;=&nbsp;Reviewed Pull Requests</span>&nbsp;&nbsp;<span>💬&nbsp;=&nbsp;Answering Questions</span>&nbsp;&nbsp;<span>🏢&nbsp;=&nbsp;Board member</span>
<table border="0">
  <tbody>
    <tr border="0">
      <td border="0" align="center" valign="top" width="16%">
        <div style="height: 100px; width: 100px;">
          <a href="https://gitlab.com/EmiliaPaz">
            <img src="https://secure.gravatar.com/avatar/c0370928d12a1dd06716ba813ce4dbcd?s=80&d=identicon" width="100px">
          </a><br>
        </div>
        <span title="Code">💻</span><br>
        <a href="https://gitlab.com/EmiliaPaz">
          <sub>Emilia Paz</sub>
        </a>
      </td>
      <td border="0" align="center" valign="top" width="16%">
        <div style="height: 100px; width: 100px;">
          <a href="https://gitlab.com/derhuerst">
            <img src="https://assets.gitlab-static.net/uploads/-/system/user/avatar/204799/avatar.png" width="100px">
          </a><br>
        </div>
        <span title="Code">💻</span><br>
        <a href="https://gitlab.com/derhuerst">
          <sub>Jannis R</sub>
        </a>
      </td>
      <td border="0" align="center" valign="top" width="16%">
        <div style="height: 100px; width: 100px;">
          <a href="https://gitlab.com/pmayd">
            <img src="https://secure.gravatar.com/avatar/aa6be35cbc4356abc6d3c13d5651110c?s=180&d=identicon" width="100px">
          </a><br>
        </div>
        <span title="Code">💻</span><br>
        <a href="https://gitlab.com/pmayd">
          <sub>Michael Aydinbas</sub>
        </a>
      </td>
      <td border="0" align="center" valign="top" width="16%">
        <div style="height: 100px; width: 100px;">
          <a href="https://gitlab.com/inktrap">
            <img src="https://secure.gravatar.com/avatar/ee9f855b89d786169f0413e76ab944e0?s=80&d=identicon" width="100px">
          </a><br>
        </div>
        <span title="Code">💻</span>&nbsp;<span title="Documentation">📝</span><br>
        <a href="https://gitlab.com/inktrap">
          <sub>Valentin</sub>
        </a>
      </td>
      <td border="0" align="center" valign="top" width="16%">
        <div style="height: 100px; width: 100px;">
          <a href="https://gitlab.com/valentin.unicorn">
            <img src="https://secure.gravatar.com/avatar/588c72c402d090166de1bd15a69fdd6b?s=80&d=identicon" width="100px">
          </a><br>
        </div>
        <span title="Code">💻</span><br>
        <a href="https://gitlab.com/valentin.unicorn">
          <sub>Valentin</sub>
        </a>
      </td>
      <td border="0" align="center" valign="top" width="16%">
        <div style="height: 100px; width: 100px;">
          <a href="https://gitlab.com/theolampert">
            <img src="https://assets.gitlab-static.net/uploads/-/system/user/avatar/2275979/avatar.png" width="100px">
          </a><br>
        </div>
        <span title="Code">💻</span><br>
        <a href="https://gitlab.com/theolampert">
          <sub>Theo</sub>
        </a>
      </td>
    </tr>
    <tr border="0">
      <td border="0" align="center" valign="top" width="16%">
        <div style="height: 100px; width: 100px;">
          <a href="https://gitlab.com/michi-zuri">
            <img src="https://assets.gitlab-static.net/uploads/-/system/user/avatar/1682847/avatar.png" width="100px">
          </a><br>
        </div>
        <span title="Bug reports">🐜</span>&nbsp;<span title="Code">💻</span>&nbsp;<span title="Design">🎨</span><br>
        <a href="https://gitlab.com/michi-zuri">
          <sub>Michael Paul Killian</sub>
        </a>
      </td>
      <td border="0" align="center" valign="top" width="16%">
        <div style="height: 100px; width: 100px;">
          <a href="https://gitlab.com/BassTii">
            <img src="https://secure.gravatar.com/avatar/f72182ecbe91d6d60603ec2c31efe7cc?s=80&d=identicon" width="100px">
          </a><br>
        </div>
        <span title="Code">💻</span><br>
        <a href="https://gitlab.com/BassTii">
          <sub>Basti A.</sub>
        </a>
      </td>
      <td border="0" align="center" valign="top" width="16%">
        <div style="height: 100px; width: 100px;">
          <a href="https://gitlab.com/chriswalg">
            <img src="https://gitlab.com/uploads/-/system/user/avatar/890523/avatar.png?width=400" width="100px">
          </a><br>
        </div>
        <span title="Bug reports">🐜</span>&nbsp;<span title="Code">💻</span>&nbsp;<span title="Documentation">📝</span><br>
        <a href="https://gitlab.com/chriswalg">
          <sub>Christian</sub>
        </a>
      </td>
      <td border="0" align="center" valign="top" width="16%">
        <div style="height: 100px; width: 100px;">
          <a href="https://gitlab.com/flukx">
            <img src="https://assets.gitlab-static.net/uploads/-/system/user/avatar/2877773/avatar.png?width=90" width="100px">
          </a><br>
        </div>
        <span title="Code">💻</span>&nbsp;<span title="Documentation">📝</span><br>
        <a href="https://gitlab.com/flukx">
          <sub> Felix</sub>
        </a>
      </td>
      <td border="0" align="center" valign="top" width="16%">
        <div style="height: 100px; width: 100px;">
          <a href="https://gitlab.com/boji.da">
            <img src="https://assets.gitlab-static.net/uploads/-/system/user/avatar/2063239/avatar.png?width=90" width="100px">
          </a><br>
        </div>
        <span title="Bug reports">🐜</span>&nbsp;<span title="Ideas, Planning, & Feedback">💡</span><br>
        <a href="https://gitlab.com/boji.da">
          <sub> Boji</sub>
        </a>
      </td>
      <td border="0" align="center" valign="top" width="16%">
        <div style="height: 100px; width: 100px;">
          <a href="https://gitlab.com/irgendwer">
            <img src="https://assets.gitlab-static.net/uploads/-/system/user/avatar/1681670/avatar.png" width="100px">
          </a><br>
        </div>
        <span title="Bug reports">🐜</span>&nbsp;<span title="Ideas, Planning, & Feedback">💡</span>&nbsp;<span title="Security">🔐</span><br>
        <a href="https://gitlab.com/irgendwer">
          <sub>irgendwer</sub>
        </a>
      </td>
    </tr>
    <tr border="0">
      <td border="0" align="center" valign="top" width="16%">
        <div style="height: 100px; width: 100px;">
          <a href="https://gitlab.com/raphaelw">
            <img src="https://avatars2.githubusercontent.com/u/7235821?s=460&v=4" width="100px">
          </a><br>
        </div>
        <span title="Code">💻</span><br>
        <a href="https://gitlab.com/raphaelw">
          <sub>Raphael</sub>
        </a>
      </td>
      <td border="0" align="center" valign="top" width="16%">
        <div style="height: 100px; width: 100px;">
          <a href="https://gitlab.com/tihar">
            <img src="https://secure.gravatar.com/avatar/ae6ead61369f0b12519b1dbbbc9c61c2?s=180&d=identicon" width="100px">
          </a><br>
        </div>
        <span title="Code">💻</span><br>
        <a href="https://gitlab.com/tihar">
          <sub>Timo Hartmann</sub>
        </a>
      </td>
      <td border="0" align="center" valign="top" width="16%">
        <div style="height: 100px; width: 100px;">
          <a href="https://gitlab.com/nicksellen">
            <img src="https://assets.gitlab-static.net/uploads/-/system/user/avatar/640443/avatar.png" width="100px">
          </a><br>
        </div>
        <span title="Code">💻</span>&nbsp;<span title="Documentation">📝</span>&nbsp;<span title="Infrastructure (Hosting, Build-Tools, etc)">🔩</span>&nbsp;<span title="Reviewed Pull Requests">👀</span><br>
        <a href="https://gitlab.com/nicksellen">
          <sub>Nick Sellen</sub>
        </a>
      </td>
      <td border="0" align="center" valign="top" width="16%">
        <div style="height: 100px; width: 100px;">
          <a href="https://gitlab.com/k.miklobusec">
            <img src="https://secure.gravatar.com/avatar/722f002e5ac2b5e0084a5c3d4ba1ce31?s=800&d=identicon" width="100px">
          </a><br>
        </div>
        <span title="Board member">🏢</span>&nbsp;<span title="Bug reports">🐜</span>&nbsp;<span title="Code">💻</span>&nbsp;<span title="Ideas, Planning, & Feedback">💡</span>&nbsp;<span title="Answering Questions">💬</span><br>
        <a href="https://gitlab.com/k.miklobusec">
          <sub>fs_k</sub>
        </a>
      </td>
      <td border="0" align="center" valign="top" width="16%">
        <div style="height: 100px; width: 100px;">
          <a href="https://gitlab.com/D0nPiano">

CHANGELOG.md
# Release "H", unreleased

## Features

## Changes
- Add E-Mail of reported user in overview !2120 @fs_k
- Store-related conversations include a link to their store now !1807 !2134 !2147 @ChrisOelmueller

## Bugfixes

## Refactoring
- move even more hardcoded language to language files !2109 @jonathan_b

## Dev/Test/CI stuff
- removed the "request" package !2136 @peter.toennies
- Upgraded deployer to the last version (v7-rc4) !2139 !2140 @\_fridtjof_
- Bump "babel" to reduce vulnerabilities !2137 @peter.toennies
- Bump "mocha" to v9 and "jsdom" to v19 to reduce vulnerabilities !2144 @peter.toennies
- bump symfony to 5.4, codeception modules to 2 !2142 @peter.toennies
- replace fzaninott/faker by fakperphp/faker !2142 @peter.toennies
- Removed old code that handled picture upload !2128 @alex.simm
- update webpack-dev-server to v4 !2130 @peter.toennies

# Release "Grapefruit", 2022-01-13

## Features

- Set iCalendar location for exported pickup events #1139 !2072 @iron9
- Display community pin for local foodsharing communities. #53 !2060 !2081 @fs_k 
- Added norwegian translation !2107 @alex.simm
- Select the community pin on a map !2102 @alex.simm 

## Changes
- Allow admins of the working group "Redaktion" to edit blog posts #34 !2061 @alex.simm
- When using a blacklisted email domain the registration process is blocked #1059 @CarolineFischer
- "New event" page requires login !2058 @Buntelrus
- Add a button to the group member list for removing members !2073 @alex.simm @chriswalg
- Add a management mode to group member list to add and remove members !2075 !2119 @chriswalg
- Add frontend functionality to display 'today' or 'tomorrow' in the event header instead of date !2069 #1120 @alexander-sav
- change german wordings in translate file !2080 @Jonathan_B
- Replace picture upload for blog posts with the new upload API #45 !2089 @alex.simm
- adds information on what gets deleted with the user profile !2106 @jonathan_b

## Bugfixes
- Fixed the list of responsible store members in REST responses !2033 #1124 @alex.simm
- Fixed for content edit !2062 @fs_k
- Fix for store change !2065 @fs_k
- Fix for future timeslot visibility to use Europe/Berlin timezone !2070 @fs_k 
- Unverified users who change their home region are not deverified again !2093 @alex.simm
- Some checks for PHP data types in order to avoid Sentry errors !2001 !2099 @alex.simm

## Refactoring
- move more hardcoded language to language files !2108 !2122 @jonathan_b
- Get Members in MemberList.vue from Rest Api !2094 !2111 !2123 @chriswalg @alex.simm
- Refactored the format of the quick search index !2115 @alex.simm 
- Get Members in MemberList.vue from Rest Api !2094 @chriswalg @alex.simm

## Dev/Test/CI stuff

- Update codeception to 4.1.22 because of a security fix !2055 @Morgy93
- Enable Gitlab dependency scanning !2056 @\_fridtjof_
- Fix the return values of database functions !2067 #916 @alex.simm
- All REST endpoints are sorted into categories !2079 #1144 @alex.simm
- exchange description for lat: lon: in translations so that it makes sense !2084 @jonathan_b
- Fix docker setup breaking down because of InfluxDB 2 !2086 @\_fridtjof_
- Add triage bot to pipeline !2074 @iron9
- Remove wrong query parameter in the calendar API #1147 @alex.simm
- added crossreferenced infos to devdocs #1149 !2082
- Run tests for websocket server ("chat server") on CI again !2091 @janopae
- exchanged name of active dev in the devdocs !2103 @Jonathan_B
- Fix warning about mkdir in upload controller !2117 @alex.simm

# Release "Feige", 2021-09-24

## Features
- Allow users to remove trust bananas on their own profile #592 !1920 @alex.simm
- Added a button that allows removing users from the email bounce list !1927 @alex.simm
- For unconfirmed pickup slots, explain in the popup text that a store manager still has to confirm it. !1949 @blinry
- Forum threads can now be closed !1851 !1990 !2045 #724 @alex.simm
- Add french and italian translation to language chooser !1964 !2039 @alex.simm
- Introduces Push Notifications for the Android App !1647 !1976 @dthulke
- Improve wording in German texts, to make the language more consistent, clear, and inclusive in some places. !1959 @blinry @Claraaa @alex.simm @fs_k
- Display same-day pickups when confirming to sign into a pickup slot !1827 !2040 @ChrisOelmueller
- Show membership in profile for workgroups !1988 !1996 @chriswalg
- IT-Support Admins can delete bananas upon request over it@foodsharing.network. !2002 @fs_k
- Added a user search field to the store management panel !2007 !2033 @alex.simm
- New calendar API including token management #80 !1719 !2029 !2045 @alex.simm
- Set iCalendar status for exported pickup events !2030 @iron9

## Changes
- Profile storelist now shows store cooperation status !1828 !1935 @ChrisOelmueller @chriswalg
- Added a partners page for foodsharing.at !1931 @alex.simm
- Redesign startpage !1778 !1982 !1991 !1997 !2015 @chriswalg @Morgy93
- Edit Team in stores is no more, functionality moved to team management mode !1810 !1811 @ChrisOelmueller

js/chat.js
var windowFocus = true;
var username;
var chatHeartbeatCount = 0;
var minChatHeartbeat = 1000;
var maxChatHeartbeat = 33000;
var chatHeartbeatTime = minChatHeartbeat;
var originalTitle;
var blinkOrder = 0;
var g_chatheartbeatTO = null;
var g_chatheartbeatTO1 = null;
var g_chatheartbeatTO2 = null;

var chatboxFocus = new Array();
var newMessages = new Array();
var newMessagesWin = new Array();
var chatBoxes = new Array();
var chatBlinkInterval = new Array();
var chatHeartbeatPause = false;
$(document).ready(function(){
	originalTitle = document.title;
	startChatSession();

	$([window, document]).blur(function(){
		windowFocus = false;
	}).focus(function(){
		windowFocus = true;
		document.title = originalTitle;
	});
});

function restructureChatBoxes() {
	align = 0;
	for (x in chatBoxes) {
		chatboxtitle = chatBoxes[x];

		if ($("#chatbox_"+chatboxtitle).css('display') != 'none') 
		{
			if (align == 0) 
			{
				$("#chatbox_"+chatboxtitle).css('right', '20px');
			} 
			else 
			{
				width = (align)*(273+7)+20;
				$("#chatbox_"+chatboxtitle).css('right', width+'px');
			}
			align++;
		}
	}
}

function chatWith(id) {
	$('.ui-dialog-content').each(function(){
		$this = $(this);

		if($this.parent().hasClass('ui-dialog'))
		{
			$this.dialog({autoOpen:false});
			$this.dialog('close');
		}
	});
	
	if ($("#chatbox_"+id).length > 0) {
		if ($("#chatbox_"+id).css('display') == 'none') {
			$("#chatbox_"+id).css('display','block');
			restructureChatBoxes();
		}
		
		$("#chatbox_"+id + ' .chatboxcontent').css('display','block');
		$("#chatbox_"+id + ' .chatboxinput').css('display','block');
		
		$("#chatbox_"+id+" .chatboxtextarea").focus();
	}
	else
	{
		var id = id;
		$.ajax({
			url: 'xhrapp.php?app=chat&m=init&id=' + id,
			dataType: 'json',
			success: function(data){
				createChatBox(data);
				if(data.msg != false)
				{
					var last_time = '';
					for(i=0;i<data.msg.length;i++)
					{
						chatAddMsg(id, data.msg[i], false);
						last_time = data.msg[i].t;
					}
					if(last_time != '')
					{
						chatAppendTime(id, last_time);
						$("#chatbox_"+id+" .chatboxcontent").scrollTop($("#chatbox_"+id+" .chatboxcontent")[0].scrollHeight);
					}
				}
				$("#chatbox_"+id+" .chatboxtextarea").focus();
			}
		});
	}
	$("#msgbar-messages").hide();
	$("#msgbar-infos").hide();
}

function killBr(str)
{
	str = str.replace('\r','');
	for(i=0;i<40;i++)
	{
		str = str.replace('\n\n\n','\n\n');
	}
	return str;
}

function chatnl2br (str, is_xhtml) {
	
	str = killBr(str).autoLink({target:'_blank'});
    var breakTag = (is_xhtml || typeof is_xhtml === 'undefined') ? '<br />' : '<br>';
    return (str + '').replace(/([^>\r\n]?)(\r\n|\n\r|\r|\n)/g, '$1' + breakTag + '$2');
}

function createChatBox(user,minimizeChatBox) {
	if ($("#chatbox_"+user.f).length > 0) {
		if ($("#chatbox_"+user.f).css('display') == 'none') {
			$("#chatbox_"+user.f).css('display','block');
			restructureChatBoxes();
		}
		$("#chatbox_"+user.f+" .chatboxtextarea").focus();
		return;
	}
	
	chatboxtitle = user.f;
	chatHiddenImg(user.f, user.p);
	var $div = $('<div />').appendTo('body');
	$div.attr("id","chatbox_"+user.f);
	$div.addClass("chatbox");
	$div.addClass("ui-corner-top");
	
	$div.html('<div class="chatboxhead ui-corner-top"><a class="chatboxtitle" href="javascript:void(0)" onclick="javascript:toggleChatBoxGrowth(\''+user.f+'\')"><i class="fa fa-comment fa-flip-horizontal"></i> '+user.n+'</a><div class="chatboxoptions"><a class="fa fa-close" href="javascript:void(0)" onclick="javascript:closeChatBox(\''+user.f+'\')"></a></div><br clear="all"/></div><div class="chatboxcontent"></div><div class="chatboxinput"><textarea placeholder="schreibe etwas..." class="chatboxtextarea" onkeydown="javascript:return checkChatBoxInputKey(event,this,\''+user.f+'\');"></textarea></div>');
	$('.chatboxinput textarea').autosize();	   
	$("#chatbox_"+user.f).css('bottom', '0px');
	
	chatBoxeslength = 0;

	for (x in chatBoxes) {
		if ($("#chatbox_"+chatBoxes[x]).css('display') != 'none') {
			chatBoxeslength++;
		}
	}

	if (chatBoxeslength == 0) {
		$("#chatbox_"+user.f).css('right', '20px');
	} else {
		width = (chatBoxeslength)*(273+7)+20;
		$("#chatbox_"+user.f).css('right', width+'px');
	}
	
	
	chatBoxes.push(user.f);
	minimize = 0;
	
	if (minimizeChatBox == 1) {
		minimizedChatBoxes = new Array();

		if ($.cookie('chatbox_minimized')) {
			minimizedChatBoxes = $.cookie('chatbox_minimized').split(/\|/);
		}
		
		for (j=0;j<minimizedChatBoxes.length;j++) {
			if (minimizedChatBoxes[j] == chatboxtitle) {
				minimize = 1;
			}
		}
	}

	chatboxFocus[chatboxtitle] = false;
	
	$("#chatbox_"+chatboxtitle+" .chatboxtextarea").blur(function(){
		chatboxFocus[chatboxtitle] = false;
		$("#chatbox_"+chatboxtitle+" .chatboxtextarea").removeClass('chatboxtextareaselected');
	}).focus(function(){
		chatboxFocus[chatboxtitle] = true;
		newMessages[chatboxtitle] = false;
		$('#chatbox_'+chatboxtitle+' .chatboxhead').removeClass('chatboxblink');
		if(chatBlinkInterval[chatboxtitle] != undefined)
		{
			clearInterval(chatBlinkInterval[chatboxtitle]);
		}
		else if(chatBlinkInterval[chatboxtitle] != false)
		{
			clearInterval(chatBlinkInterval[chatboxtitle]);
		}
		chatBlinkInterval[chatboxtitle] = false;

	});

	$("#chatbox_"+chatboxtitle).click(function() {
		if ($('#chatbox_'+chatboxtitle+' .chatboxcontent').css('display') != 'none') {
			//$("#chatbox_"+chatboxtitle+" .chatboxtextarea").focus();
		}
	});

	$("#chatbox_"+chatboxtitle).show();
	
	$('.chatboxtextarea').focus(function(){
		$('.chatboxtextarea').removeClass('chatboxtextareaselected');
		$(this).addClass('chatboxtextareaselected');
	});
	
	$('.chatboxtextarea').keydown(function(objEvent) {
	    if (objEvent.keyCode == 9) { 
	        objEvent.preventDefault();
	    }
	});
	
	$("#chatbox_"+chatboxtitle+" .chatboxcontent").slimScroll({
        height: '264px'
    });
	$('#chatbox_'+chatboxtitle+' .chatboxcontent').css('height','250px');
	
	if (minimize == 1) {
		$('#chatbox_'+chatboxtitle+' .chatboxcontent').css('display','none');
		$('#chatbox_'+chatboxtitle+' .slimScrollDiv').css('display','none');
		$('#chatbox_'+chatboxtitle+' .chatboxinput').css('display','none');
	}
}

function stopChatHeartbeat()
{
	chatHeartbeatPause = true;
}

function chatHeartbeat(){

	
	if(chatHeartbeatPause)
	{
		return false;
	}
	
	var itemsfound = 0;
	
	if (windowFocus == false) {
 
		var blinkNumber = 0;
		var titleChanged = 0;
		for (x in newMessagesWin) {
			if (newMessagesWin[x] == true) {
				++blinkNumber;
				if (blinkNumber >= blinkOrder) {
					document.title = 'Neue Nachricht...';
					titleChanged = 1;
					break;	
				}
			}
		}
		
		if (titleChanged == 0) {
			document.title = originalTitle;
			blinkOrder = 0;
		} else {
			++blinkOrder;
		}

	} else {
		for (x in newMessagesWin) {
			newMessagesWin[x] = false;
		}
	}

	for (x in newMessages) {
		if (newMessages[x] == true) {
			if (chatboxFocus[x] == false) {
				//FIXME: add toggle all or none policy, otherwise it looks funny
				if(chatBlinkInterval[x] == undefined || chatBlinkInterval[x] == false)
				{
					chatBlinkInterval[x] = setInterval(function(){
						$('#chatbox_'+x+' .chatboxhead').toggleClass('chatboxblink');
					},1000);
				}
			}
		}
	}
	
	$.ajax({
	  url: "chat.php?action=chatheartbeat",
	  cache: false,
	  dataType: "json",
	  success: function(data) {
		
		$.each(data.items, function(i,item){
			if (item)	{ // fix strange ie bug
				chatAddMsg(item.f,item);
				itemsfound += 1;
				last_time = item.t;
			}
		});

		chatHeartbeatCount++;

		if (itemsfound > 0) {
			chatHeartbeatTime = minChatHeartbeat;
			chatHeartbeatCount = 1;
		} else if (chatHeartbeatCount >= 10) {
			chatHeartbeatTime *= 2;
			chatHeartbeatCount = 1;
			if (chatHeartbeatTime > maxChatHeartbeat) {
				chatHeartbeatTime = maxChatHeartbeat;
			}
		}
		
		g_chatheartbeatTO = setTimeout('chatHeartbeat();',chatHeartbeatTime);
	}});
}

function chatAddMsg(id,item,newmsg)
{
	if(newmsg == undefined)
	{
		newmsg = true;
	}
	chatboxtitle = id;
	name = item.n;

	if ($("#chatbox_"+chatboxtitle).length <= 0) {
		createChatBox(item);
	}
	if ($("#chatbox_"+chatboxtitle).css('display') == 'none') {
		$("#chatbox_"+chatboxtitle).css('display','block');
		restructureChatBoxes();
	}
	
	if (item.s == 1) {
		item.f = username;
	}

	if (item.s == 2) {
		chatAppendTime(id, item.m);
		
	} else {
		if(newmsg)
		{
			newMessages[chatboxtitle] = true;
			newMessagesWin[chatboxtitle] = true;
		}
		$("#chatbox_"+chatboxtitle+" .chatboxcontent").append('<div class="chatboxmessage" title="'+item.sent+'"><span class="chatboxmessagefrom">'+chatImg(item.f)+'</span><span class="chatboxmessagecontent">'+nl2br(item.m)+'</span><div style="clear:both;"></div></div>');
	}
	if($("#chatbox_"+chatboxtitle+" .chatboxcontent").length > 0)
	{
		$("#chatbox_"+chatboxtitle+" .chatboxcontent").scrollTop($("#chatbox_"+chatboxtitle+" .chatboxcontent")[0].scrollHeight);
	}
}

function chatAppendTime(id,time)
{
	if(time != undefined)
	{
		$("#chatbox_"+id+" .chatboxcontent").append('<div class="chatboxmessage"><span class="chatboxinfo" title="'+time+'">'+time+'</span></div>');
		$('.chatboxinfo').timeago();
	}
}

function closeChatBox(chatboxtitle) {
	$('#chatbox_'+chatboxtitle).css('display','none');
	restructureChatBoxes();

	$.post("chat.php?action=closechat", { chatbox: chatboxtitle} , function(data){	
	});

}

function toggleChatBoxGrowth(chatboxtitle) {
	if ($('#chatbox_'+chatboxtitle+' .chatboxcontent').css('display') == 'none') {  
		
		var minimizedChatBoxes = new Array();
		
		if ($.cookie('chatbox_minimized')) {
			minimizedChatBoxes = $.cookie('chatbox_minimized').split(/\|/);
		}

		var newCookie = '';

		for (i=0;i<minimizedChatBoxes.length;i++) {
			if (minimizedChatBoxes[i] != chatboxtitle) {
				newCookie += chatboxtitle+'|';
			}
		}

		newCookie = newCookie.slice(0, -1)


		$.cookie('chatbox_minimized', newCookie);
		$('#chatbox_'+chatboxtitle+' .chatboxcontent').css('display','block');
		$('#chatbox_'+chatboxtitle+' .slimScrollDiv').css('display','block');
		//slimScrollDiv
		$('#chatbox_'+chatboxtitle+' .chatboxinput').css('display','block');
		$("#chatbox_"+chatboxtitle+" .chatboxcontent").scrollTop($("#chatbox_"+chatboxtitle+" .chatboxcontent")[0].scrollHeight);
		
		
	} else {
		
		var newCookie = chatboxtitle;

		if ($.cookie('chatbox_minimized')) {
			newCookie += '|'+$.cookie('chatbox_minimized');
		}


		$.cookie('chatbox_minimized',newCookie);
		$('#chatbox_'+chatboxtitle+' .chatboxcontent').css('display','none');
		$('#chatbox_'+chatboxtitle+' .slimScrollDiv').css('display','none');
		$('#chatbox_'+chatboxtitle+' .chatboxinput').css('display','none');
	}
	
}

function checkChatBoxInputKey(event,chatboxtextarea,chatboxtitle) {
	if(event.keyCode == 13 && event.shiftKey == 0)  {
		message = $(chatboxtextarea).val();
		message = message.replace(/^\s+|\s+$/g,"");
		killBr(message);
		$(chatboxtextarea).val('');
		//$(chatboxtextarea).focus();
		$(chatboxtextarea).css('height','44px');
		if (message != '') {
			$.ajax({
				url: "xhrapp.php?app=chat&m=sendchat",
				data: {to: chatboxtitle, message: message},
				type: 'post',
				dataType: 'json',
				success: function(data){
					message = message.replace(/</g,"&lt;").replace(/>/g,"&gt;").replace(/\"/g,"&quot;");

					$("#chatbox_"+chatboxtitle+" .chatboxcontent").append('<div class="chatboxmessage"><span class="chatboxmessagefrom">'+chatImg(username)+'</span><span class="chatboxmessagecontent">'+chatnl2br(message)+'</span><div style="clear:both;"></div></div>');
					$("#chatbox_"+chatboxtitle+" .chatboxcontent").scrollTop($("#chatbox_"+chatboxtitle+" .chatboxcontent")[0].scrollHeight);
				}
			});
			
		}
		chatHeartbeatTime = minChatHeartbeat;
		chatHeartbeatCount = 1;

		return false;
	}

	var adjustedHeight = chatboxtextarea.clientHeight;
	var maxHeight = 94;

	if (maxHeight > adjustedHeight) {
		adjustedHeight = Math.max(chatboxtextarea.scrollHeight, adjustedHeight);
		if (maxHeight)
			adjustedHeight = Math.min(maxHeight, adjustedHeight);
		if (adjustedHeight > chatboxtextarea.clientHeight)
			$(chatboxtextarea).css('height',adjustedHeight+8 +'px');
	} else {
		$(chatboxtextarea).css('overflow','auto');
	}
	 
}

function chatHiddenImg(id,photo)
{
	if(photo != undefined && photo != '')
	{
		$('#chatHidden .hImg-'+id).remove();
		photo = '<a onclick="profile('+parseInt(id)+');return false;" class="photo" href="#'+id+'"><img src="images/mini_q_'+photo+'" /></a>';
		$('#chatHidden').append('<div class="hImg-'+id+'">'+photo+'</div>');
	}
}

function startChatSession(){  
	$('body:last').append('<div id="chatHidden" style="display:none;"></div>');
	$.ajax({
	  url: "chat.php?action=startchatsession",
	  cache: false,
	  dataType: "json",
	  success: function(data) {
 
		username = data.username;
		chat_photo = data.photo;
		chatHiddenImg(username,data.photo);
		$.each(data.items, function(i,item){
			if (item)	{ // fix strange ie bug

				chatboxtitle = item.f;

				if ($("#chatbox_"+chatboxtitle).length <= 0) {
					createChatBox(item,1);
				}
				
				if (item.s == 1) {
					item.f = username;
				}

				if (item.s == 2) {
					$("#chatbox_"+chatboxtitle+" .chatboxcontent").append('<div class="chatboxmessage"><span class="chatboxinfo" title="'+item.m+'">'+item.m+'</span></div>');
					$('.chatboxinfo').timeago();
				} else {
					$("#chatbox_"+chatboxtitle+" .chatboxcontent").append('<div class="chatboxmessage"><span class="chatboxmessagefrom">'+chatImg(item.f)+'</span><span class="chatboxmessagecontent">'+chatnl2br(item.m)+'</span><div style="clear:both;"></div></div>');
				}
			}
		});
		
		for (i=0;i<chatBoxes.length;i++) {
			chatboxtitle = chatBoxes[i];
			$("#chatbox_"+chatboxtitle+" .chatboxcontent").scrollTop($("#chatbox_"+chatboxtitle+" .chatboxcontent")[0].scrollHeight);
			g_chatheartbeatTO1 = setTimeout('$("#chatbox_"+chatboxtitle+" .chatboxcontent").scrollTop($("#chatbox_"+chatboxtitle+" .chatboxcontent")[0].scrollHeight);', 100); // yet another strange ie bug
		}
	
		g_chatheartbeatTO2 = setTimeout('chatHeartbeat();',chatHeartbeatTime);
		
	}});
}

function chatImg(id)
{
	if($('#chatHidden .hImg-'+id).length > 0)
	{
		return $('#chatHidden .hImg-'+id).html();
	}
	else
	{
		return '';
		
	}
}

/**
 * Cookie plugin
 *
 * Copyright (c) 2006 Klaus Hartl (stilbuero.de)
 * Dual licensed under the MIT and GPL licenses:
 * http://www.opensource.org/licenses/mit-license.php
 * http://www.gnu.org/licenses/gpl.html
 *
 */

jQuery.cookie = function(name, value, options) {
    if (typeof value != 'undefined') { // name and value given, set cookie
        options = options || {};
        if (value === null) {
            value = '';
            options.expires = -1;
        }
        var expires = '';
        if (options.expires && (typeof options.expires == 'number' || options.expires.toUTCString)) {
            var date;
            if (typeof options.expires == 'number') {
                date = new Date();
                date.setTime(date.getTime() + (options.expires * 24 * 60 * 60 * 1000));
            } else {
                date = options.expires;
            }
            expires = '; expires=' + date.toUTCString(); // use expires attribute, max-age is not supported by IE
        }
        // CAUTION: Needed to parenthesize options.path and options.domain
        // in the following expressions, otherwise they evaluate to undefined
        // in the packed version for some reason...
        var path = options.path ? '; path=' + (options.path) : '';
        var domain = options.domain ? '; domain=' + (options.domain) : '';
        var secure = options.secure ? '; secure' : '';
        document.cookie = [name, '=', encodeURIComponent(value), expires, path, domain, secure].join('');
    } else { // only name given, get cookie
        var cookieValue = null;
        if (document.cookie && document.cookie != '') {
            var cookies = document.cookie.split(';');
            for (var i = 0; i < cookies.length; i++) {
                var cookie = jQuery.trim(cookies[i]);
                // Does this cookie string begin with the name we want?
                if (cookie.substring(0, name.length + 1) == (name + '=')) {
                    cookieValue = decodeURIComponent(cookie.substring(name.length + 1));
                    break;
                }
            }
        }
        return cookieValue;
    }
};
